crates/wit-parser/src/resolve.rs

```diff
@@ -293,32 +309,64 @@ impl Resolve {
     /// The second value returned here is the list of paths that were parsed
     /// when generating the return value. This can be useful for build systems
     /// that want to rebuild bindings whenever one of the files change.
-    pub fn push_dir(&mut self, path: impl AsRef<Path>) -> Result<(Vec<PackageId>, Vec<PathBuf>)> {
-        let path = path.as_ref();
-        let deps_path = path.join("deps");
-        let unresolved_deps = self.parse_deps_dir(&deps_path).with_context(|| {
-            format!(
-                "failed to parse dependency directory: {}",
-                deps_path.display()
-            )
-        })?;
-        let (_, mut path_bufs) = self.sort_unresolved_packages(unresolved_deps)?;
-
-        let unresolved_top_level = UnresolvedPackageGroup::parse_dir(path)
+    pub fn push_dir(&mut self, path: &Path) -> Result<(PackageId, Vec<PathBuf>)> {
+        let top_pkg = UnresolvedPackageGroup::parse_dir(path)
             .with_context(|| format!("failed to parse package: {}", path.display()))?;
-        let (pkgs_ids, mut top_level_path_bufs) =
-            self.sort_unresolved_packages(vec![unresolved_top_level])?;
+        let deps = path.join("deps");
+        let mut deps = self
+            .parse_deps_dir(&deps)
+            .with_context(|| format!("failed to parse dependency directory: {}", deps.display()))?;
+        // Perform a simple topological sort which will bail out on cycles
+        // and otherwise determine the order that packages must be added to
+        // this `Resolve`.
+        let mut order = IndexSet::new();
+        let mut visiting = HashSet::new();
+        for pkg in &top_pkg.nested {
+            visit_group(&pkg, &deps, &mut order, &mut visiting, &pkg.source_map)?;
+        }
+        for pkg in deps.values().chain([&top_pkg]) {
+            for pkg in &pkg.nested {
+                visit_group(&pkg, &deps, &mut order, &mut visiting, &pkg.source_map)?;
+            }
+            visit_group(&pkg.main, &deps, &mut order, &mut visiting, &pkg.source_map)?;
+        }
```

Comment:
Is it possibly to rely on `sort_unresolved_packages` here instead of duplicating some parts of it? Ideally there'd be just one "sort this soup" method but I also realize that there's a lot of entry points here and may make that complicated to write.